File: backend/src/routes/tmdb.rs
use crate::routes::{with_form_body, respond};
use crate::utils::error_handling::{AppError, ErrorType};

use warp::Filter;
use serde::{Serialize, Deserialize};
use std::env;


// MOVIE DETAILS
// **************************************************

// REQUEST PARAMS
#[derive(Deserialize)]
struct MovieDetailsParams {
  movie_id: String
}

// RESPONSE DATA
#[derive(Serialize, Deserialize)]
struct Genre {
  id: Option<i32>,
  name: Option<String>
}

#[derive(Serialize, Deserialize)]
struct ProductionCompany {
  id: Option<i32>,
  logo_path: Option<String>,
  name: Option<String>,
  origin_country: Option<String>
}

#[derive(Serialize, Deserialize)]
struct ProductionCountry {
  iso_3166_1: Option<String>,
  name: Option<String>,
}

#[derive(Serialize, Deserialize)]
struct Language {
  english_name: Option<String>,
  iso_639_1: Option<String>,
  name: Option<String>
}

// VIDEOS
#[derive(Serialize, Deserialize)]
struct Video {
  iso_639_1: Option<String>,
  iso_3166_1: Option<String>,
  name: Option<String>,
  key: Option<String>,
  site: Option<String>,
  size: Option<i32>,
  r#type: Option<String>,
  official: Option<bool>,
  published_at: Option<String>,
  id: Option<String>
}

#[derive(Serialize, Deserialize)]
struct Videos {
  id: Option<i32>,
  results: Option<Box<[Video]>>
}
// IMAGES
#[derive(Serialize, Deserialize)]
struct Image {
  aspect_ration: Option<f32>,
  height: Option<i32>,
  iso_639_1: Option<String>,
  file_path: Option<String>,
  vote_average: Option<f32>,
  vote_count: Option<i32>,
  width: Option<i32>
}

#[derive(Serialize, Deserialize)]
struct Images {
  id: Option<i32>,
  backdrops: Option<Box<[Image]>>,
  logos: Option<Box<[Image]>>,
  // posters: Option<Box<[Image]>>
}

// CREDITS
#[derive(Serialize, Deserialize)]
struct CastCrewMember {

  // cast & crew
  adult: Option<bool>,
  gender: Option<i32>,
  id: Option<i32>,
  known_for_department: Option<String>,
  name: Option<String>,
  original_name: Option<String>,
  popularity: Option<f32>,
  profile_path: Option<String>,
  credit_id: Option<String>,

  // cast
  cast_id: Option<i32>,
  character: Option<String>,
  order: Option<i32>,

  // crew
  department: Option<String>,
  job: Option<String>,
}

#[derive(Serialize, Deserialize)]
struct Credits {
  id: Option<i32>,
  cast: Option<Box<[CastCrewMember]>>,
  crew: Option<Box<[CastCrewMember]>>
}

// MAIN RESPONSE
#[derive(Serialize, Deserialize)]
struct MovieDetails {
  adult: Option<bool>,
  backdrop_path: Option<String>,
  // belongs_to_collection: Option<String>,
  budget: Option<i32>,
  genres: Option<Box<[Genre]>>,
  homepage: Option<String>,
  id: Option<i32>,
  imdb_id: Option<String>,
  original_language: Option<String>,
  original_title: Option<String>,
  overview: Option<String>,
  popularity: Option<f32>,
  poster_path: Option<String>,
  production_companies: Option<Box<[ProductionCompany]>>,
  production_countries: Option<Box<[ProductionCountry]>>,
  release_date: Option<String>,
  revenue: Option<i32>,
  runtime: Option<i32>,
  spoken_languages: Option<Box<[Language]>>,
  status: Option<String>,
  tagline: Option<String>,
  title: Option<String>,
  video: Option<bool>,
  vote_average: Option<f32>,
  vote_count: Option<i32>,
  videos: Option<Videos>,
  images: Option<Images>,
  credits: Option<Credits>
}

// PERSON DETAILS
// **************************************************
// REQUEST PARAMS
#[derive(Deserialize)]
struct PersonDetailsParams {
  person_id: String
}

#[derive(Serialize, Deserialize)]
struct PersonCredits {
  adult: Option<bool>,
  backdrop_path: Option<String>,
  genre_ids: Option<Box<[i32]>>,
  id: Option<i32>,
  original_language: Option<String>,
  original_title: Option<String>,
  overview: Option<String>,
  popularity: Option<f32>,
  poster_path: Option<String>,
  release_date: Option<String>,
  title: Option<String>,
  video: Option<bool>,
  vote_average: Option<f32>,
  vote_count: Option<i32>,
  credit_id: Option<String>,

  // cast
  character: Option<String>,
  order: Option<i32>,

  // crew
  department: Option<String>,
  job: Option<String>
}

// MAIN RESPONSE
#[derive(Serialize, Deserialize)]
struct PersonDetails {
  cast: Option<Box<[PersonCredits]>>,
  crew: Option<Box<[PersonCredits]>>
}

// SEARCH
// **************************************************

// REQUEST PARAMS
#[derive(Deserialize)]
struct QueryParams {
  query: String,
  page: String,
  endpoint: String
}

// RESPONSE DATA
#[derive(Serialize, Deserialize)]
struct KnownFor {
  adult: Option<bool>,
  backdrop_path: Option<String>,
  id: Option<i32>,
  title: Option<String>,
  original_language: Option<String>,
  original_title: Option<String>,
  overview: Option<String>,
  poster_path: Option<String>,
  media_type: Option<String>,
  genre_ids: Option<Box<[i32]>>,
  popularity: Option<f32>,
  release_date: Option<String>,
  video: Option<bool>,
  vote_average: Option<f32>,
  vote_count: Option<i32>
}

#[derive(Serialize, Deserialize)]
struct SearchResult {
  adult: Option<bool>,
  backdrop_path: Option<String>,
  genre_ids: Option<Box<[i32]>>,
  id: Option<i32>,
  original_language: Option<String>,
  original_title: Option<String>,
  overview: Option<String>,
  popularity: Option<f32>,
  poster_path: Option<String>,
  release_date: Option<String>,
  title: Option<String>,
  video: Option<bool>,
  vote_average: Option<f32>,
  vote_count: Option<i32>,

  gender: Option<i32>,
  known_for_department: Option<String>,
  name: Option<String>,
  original_name: Option<String>,
  profile_path: Option<String>,
  known_for: Option<Box<[KnownFor]>>,

  media_type: Option<String>,
}

// MAIN RESPONSE
#[derive(Serialize, Deserialize)]
struct SearchResults {
  page: Option<i32>,
  results: Option<Box<[SearchResult]>>,
  total_pages: Option<i32>,
  total_results: Option<i32>,
}


// ENDPOINTS
// *******************************
pub fn tmdb_filters()
-> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone
{
  search_filters()
  .or(movie_details_filters())
  .or(person_details_filters())
}

pub fn search_filters()
-> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone
{
  warp::path!("tmdb" / "search")
    .and(warp::post())
    .and(with_form_body::<QueryParams>())
    .and_then(search)
}

async fn search(query_params: QueryParams)
-> Result<impl warp::Reply, warp::Rejection>
{
  let tmdb_base_url = env::var("TMDB_BASE_URL").unwrap();
  let tmdb_api_key = env::var("TMDB_API_KEY").unwrap();

  let search_url = format!("{}/search/{}?api_key={}&query={}&page={}",
    &tmdb_base_url,
    &query_params.endpoint,
    &tmdb_api_key,
    &query_params.query,
    &query_params.page);

  let response = reqwest::get(&search_url).await.unwrap()
    .json::<SearchResults>().await
    .map_err(|err| {
      AppError::new(&err.to_string(), ErrorType::FailedToSearch)
    });

  respond(response, warp::http::StatusCode::OK)
}

pub fn movie_details_filters()
-> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone
{
  warp::path!("tmdb" / "movie")
    .and(warp::post())
    .and(with_form_body::<MovieDetailsParams>())
    .and_then(movie_details)
}

async fn movie_details(movie_details_params: MovieDetailsParams)
-> Result<impl warp::Reply, warp::Rejection>
{
  let tmdb_base_url = env::var("TMDB_BASE_URL").unwrap();
  let tmdb_api_key = env::var("TMDB_API_KEY").unwrap();

  let movie_details_url = format!("{}/movie/{}?api_key={}&language=en-US&append_to_response=videos,images,credits&include_image_language=en,null",
    &tmdb_base_url,
    &movie_details_params.movie_id,
    &tmdb_api_key);

    let response = reqwest::get(&movie_details_url).await.unwrap()
    .json::<MovieDetails>().await
    .map_err(|err| {
      AppError::new(&err.to_string(), ErrorType::FailedToGetMovieDetails)
    });

  respond(response, warp::http::StatusCode::OK)
}

fn person_details_filters()
-> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone
{
  warp::path!("tmdb" / "person")
    .and(warp::post())
    .and(with_form_body::<PersonDetailsParams>())
    .and_then(person_details)
}

async fn person_details(person_details_params: PersonDetailsParams)
-> Result<impl warp::Reply, warp::Rejection>
{
  let tmdb_base_url = env::var("TMDB_BASE_URL").unwrap();
  let tmdb_api_key = env::var("TMDB_API_KEY").unwrap();

  let person_details_url = format!("{}/person/{}/movie_credits?api_key={}",
    &tmdb_base_url,
    &person_details_params.person_id,
    &tmdb_api_key);
  
  let response = reqwest::get(&person_details_url).await.unwrap()
    .json::<PersonDetails>().await
    .map_err(|err| {
      AppError::new(&err.to_string(), ErrorType::FailedToGetPersonDetails)
    });
  
  respond(response, warp::http::StatusCode::OK)
}

File: v1/movieshlok/frontend/redux/reducers/tmdb.ts
import { FilterResults } from "@/pages/search";
import {
  getSearchResults,
  getMovieDetails,
  getPersonDetails,
  PersonData
} from "@/redux/actions/tmdb";
import { Status } from "@/redux/reducers/index";
import { AppState } from "@/redux/store";
import { createSlice } from "@reduxjs/toolkit";

// MOVIE DETAILS
// **************************************************

// RESPONSE DATA
export interface Genre {
  id?: number;
  name?: string;
}

export interface ProductionCompany {
  id?: number;
  logo_path?: string;
  name?: string;
  origin_country?: string;
}

export interface ProductionCountry {
  iso_3166_1?: string;
  name?: string;
}

export interface Language {
  english_name?: string;
  iso_639_1?: string;
  name?: string;
}

// VIDEOS
export interface Video {
  iso_639_1?: string;
  iso_3166_1?: string;
  name?: string;
  key?: string;
  site?: string;
  size?: number;
  type?: string;
  official?: boolean;
  published_at?: string;
  id?: string;
}

export interface Videos {
  id?: number;
  results?: [Video];
}

// IMAGES
export interface MovieImage {
  aspect_ration?: number;
  height?: number;
  iso_639_1?: string;
  file_path?: string;
  vote_average?: number;
  vote_count?: number;
  width?: number;
}

interface Images {
  id?: number;
  backdrops?: [MovieImage];
  logos?: [MovieImage];
}

// CREDITS
export interface CastCrewMember {
  // cast & crew
  adult?: boolean;
  gener?: number;
  id: number;
  known_for_department?: string;
  name?: string;
  original_name?: string;
  popularity?: number;
  profile_path?: string;
  credit_id?: string;

  // cast
  cast_id?: number;
  character?: string;
  order?: number;

  // crew
  department?: string;
  job?: string;
}

interface Credits {
  id?: number;
  cast?: CastCrewMember[];
  crew?: CastCrewMember[];
}

// MAIN RESPONSE
export interface MovieDetails {
  status: Status;
  success: boolean | null;
  message: string;
  data: {
    adult?: boolean;
    backdrop_path?: string;
    budget?: number;
    genres?: [Genre];
    homepage?: string;
    id?: number;
    imdb_id?: string;
    original_language?: string;
    original_title?: string;
    overview?: string;
    popularity?: number;
    poster_path?: string;
    production_companies?: [ProductionCompany];
    production_countries?: [ProductionCountry];
    release_date?: string;
    revenue?: number;
    runtime?: number;
    spoken_languages?: [Language];
    status?: string;
    tagline?: string;
    title?: string;
    video?: boolean;
    vote_average?: number;
    vote_count?: number;
    videos?: Videos;
    images?: Images;
    credits?: Credits;
  };
}

// INITIAL STATE
const initialMovieDetailsState: MovieDetails = {
  status: "idle",
  success: null,
  message: "",
  data: {}
};

// SEARCH
// **************************************************

// RESPONSE DATA
export interface KnownFor {
  id: number;
  title: string;
  media_type: string;
}

export interface SearchResult {
  id: number;
  popularity: number;
  media_type?: string; // only multi
  title?: string; // only movie
  poster_path?: string; // only movie
  overview?: string; // only movie
  release_date?: string; // only movie
  vote_average?: number; // only movie
  vote_count?: number; // only movie
  name?: string; // only person
  profile_path?: string; // only person
  known_for?: KnownFor[]; // only person
  known_for_department?: string;
}

// MAIN RESPONSE
interface SearchResults {
  status: Status;
  success: boolean | null;
  message: string;
  query: string;
  filter: FilterResults | null;
  data: {
    page?: string;
    results?: [SearchResult];
    total_pages?: string;
  };
}

// INITIAL STATE
const initialSearchResultsState: SearchResults = {
  status: "idle",
  success: null,
  message: "",
  query: "",
  filter: null,
  data: {}
};

// PERSON DETAILS
// ****************************
interface PersonDetails {
  status: Status;
  success: boolean | null;
  message: string;
  details: PersonData | null;
}

const initialPersonDetailsState: PersonDetails = {
  status: "idle",
  success: null,
  message: "",
  details: null
};

export const searchResultsSlice = createSlice({
  name: "searchResults",
  initialState: initialSearchResultsState,
  reducers: {},
  extraReducers: builder => {
    builder
      .addCase(getSearchResults.pending, state => {
        state.status = "loading";
        state.success = null;
        state.message = "";
        state.query = "";
        state.filter = null;
        state.data = {};
      })
      .addCase(getSearchResults.fulfilled, (state, action) => {
        state.status = "fulfilled";
        state.success = action.payload.success;
        state.message = action.payload.message;
        state.query = action.payload.query;
        state.filter = action.payload.filter;
        state.data = action.payload.data;
      });
  }
});

export const selectSearchResults = (state: AppState) => state.searchResults;
export const searchResultsReducer = searchResultsSlice.reducer;

export const movieDetailsSlice = createSlice({
  name: "movieDetails",
  initialState: initialMovieDetailsState,
  reducers: {},
  extraReducers: builder => {
    builder
      .addCase(getMovieDetails.pending, state => {
        state.status = "loading";
        state.success = null;
        state.message = "";
        state.data = {};
      })
      .addCase(getMovieDetails.fulfilled, (state, action) => {
        state.status = "fulfilled";
        state.success = action.payload.success;
        state.message = action.payload.message;
        state.data = action.payload.data;
      });
  }
});

export const selectMovieDetails = (state: AppState) => state.movieDetails;
export const movieDetailsReducer = movieDetailsSlice.reducer;

export const personDetailsSlice = createSlice({
  name: "personDetails",
  initialState: initialPersonDetailsState,
  reducers: {},
  extraReducers: builder => {
    builder
      .addCase(getPersonDetails.pending, state => {
        state.status = "loading";
        state.success = null;
        state.message = "";
        state.details = null;
      })
      .addCase(getPersonDetails.fulfilled, (state, action) => {
        state.status = "fulfilled";
        state.success = action.payload.success;
        state.message = action.payload.message;
        state.details = action.payload.details;
      });
  }
});

export const selectPersonDetails = (state: AppState) => state.personDetails;
export const personDetailsReducer = personDetailsSlice.reducer;
